In [1]:
import pandas as pd

In [2]:
dt = pd.read_csv("../Dataset/clean_medicine_data.csv")

In [3]:
dt.head(5)

,id,name,price,Is_discontinued,manufacturer_name,type,pack_size_label,medicine_desc,side_effects,drug_interactions,composition,cleaned_name,document
0,1,Augmentin 625 Duo Tablet,223,False,glaxo smithkline pharmaceuticals ltd,allopathy,strip of 10 tablets,Augmentin 625 Duo Tablet is a penicillin-type ...,"Vomiting,Nausea,Diarrhea","{""drug"": [], ""brand"": [], ""effect"": []}",amoxycillin (500mg) clavulanic acid (125mg),augmentin 625 duo tablet,"\n Medicine : augmentin 625 duo tablet,\n ..."
1,2,Azithral 500 Tablet,132,False,alembic pharmaceuticals ltd,allopathy,strip of 5 tablets,Azithral 500 Tablet is an antibiotic used to t...,"Vomiting,Nausea,Abdominal pain,Diarrhea","{""drug"": [""Astemizole"", ""Ebastine"", ""Mizolasti...",azithromycin (500mg),azithral 500 tablet,"\n Medicine : azithral 500 tablet,\n Com..."
2,3,Ascoril LS Syrup,118,False,glenmark pharmaceuticals ltd,allopathy,bottle of 100 ml syrup,Ascoril LS Syrup is a combination medicine use...,"Nausea,Vomiting,Diarrhea,Upset stomach,Stomach...","{""drug"": [], ""brand"": [], ""effect"": []}",ambroxol (30mg/5ml) levosalbutamol (1mg/5ml),ascoril ls syrup,"\n Medicine : ascoril ls syrup,\n Compos..."
3,4,Allegra 120mg Tablet,218,False,sanofi india ltd,allopathy,strip of 10 tablets,Allegra 120mg Tablet is an anti-allergy medici...,"Headache,Drowsiness,Dizziness,Nausea","{""drug"": [""Aprepitant"", ""Fosaprepitant"", ""Maga...",fexofenadine (120mg),allegra 120mg tablet,"\n Medicine : allegra 120mg tablet,\n Co..."
4,5,Avil 25 Tablet,10,False,sanofi india ltd,allopathy,strip of 15 tablets,Avil 25 Tablet is an antiallergic medication u...,Sleepiness,"{""drug"": [""Zafirlukast"", ""Alprazolam"", ""Betahi...",pheniramine (25mg),avil 25 tablet,"\n Medicine : avil 25 tablet,\n Composit..."


#### We will create Dictionaries to map name to index and vice-versa for fast lookup

In [4]:
name_to_index = {
    name:idx for idx,name in enumerate(dt['cleaned_name'])
}

index_to_name = {
    idx:name for idx,name in enumerate(dt['cleaned_name'])
}

In [2]:
def get_composition(medicine:str):
    medicine = medicine.lower().strip()

    if medicine not in name_to_index:
        return None
    
    idx = name_to_index[medicine]
    return dt.loc[idx, 'composition']

In [3]:
def find_exact_substitutes(medicine:str, top_k=5):
    comp = get_composition(medicine)

    if comp is None:
        return "Medicine not found."

    substitutes = dt[dt['composition'] == comp]
    substitutes = substitutes.sort_values("price")

    return substitutes[[
        "name",
        "manufacturer_name",
        "price",
        "pack_size_label",
        "composition"
    ]].head(top_k)

In [7]:
find_exact_substitutes("Allegra 120mg Tablet")

,name,manufacturer_name,price,pack_size_label,composition
87339,Faxodas 120mg Tablet,shreya life sciences pvt ltd,19,strip of 6 tablets,fexofenadine (120mg)
236774,Zelgra 120 Tablet,zeelab pharmacy pvt ltd,22,strip of 10 tablets,fexofenadine (120mg)
1265,Alert 120mg Tablet,alkem laboratories ltd,28,strip of 10 tablets,fexofenadine (120mg)
60041,DavaIndia Fexofenadine 120mg Tablet,davaindia generic pharmacy,29,strip of 10 tablets,fexofenadine (120mg)
4458,Allegix 120mg Tablet,intas pharmaceuticals ltd,30,strip of 6 tablets,fexofenadine (120mg)


In [8]:
find_exact_substitutes("benadryl syrup")

,name,manufacturer_name,price,pack_size_label,composition
103810,Intadryl Syrup,intas pharmaceuticals ltd,38,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...
156341,Odaril Syrup,moraceae pharmaceuticals pvt ltd,43,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...
207575,Tussberry-N Syrup,acinta pharmaceuticals pvt ltd,51,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...
145633,New Brethese Syrup,galpha laboratories ltd,68,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...
193007,Suntuss CF Syrup,suncare formulations pvt ltd,70,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...


### Similarity Matching (TF-IDF)

In [4]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

tfidf = TfidfVectorizer()
comp_vectors = tfidf.fit_transform(dt['composition'])

NameError: name 'dt' is not defined

In [ ]:
def find_similar_substitutes(medicine:str, top_k=5):
    idx = name_to_index.get(medicine.lower())

    if idx is None:
        return "Medicine not found."
    
    query_vec = comp_vectors[idx]

    similarities = cosine_similarity(query_vec, comp_vectors)[0]

    dt["similarity"] = similarities

    result = dt.sort_values(['similarity', 'price'],ascending=[False,True])

    return result[[
        "name",
        "manufacturer_name",
        "price",
        "pack_size_label",
        "composition",
        "similarity"
    ]].head(20)

In [11]:
find_similar_substitutes("benadryl syrup")

,name,manufacturer_name,price,pack_size_label,composition,similarity
103810,Intadryl Syrup,intas pharmaceuticals ltd,38,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,1.000000
156341,Odaril Syrup,moraceae pharmaceuticals pvt ltd,43,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,1.000000
207575,Tussberry-N Syrup,acinta pharmaceuticals pvt ltd,51,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,1.000000
145633,New Brethese Syrup,galpha laboratories ltd,68,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,1.000000
193007,Suntuss CF Syrup,suncare formulations pvt ltd,70,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,1.000000
197356,Super Tus Syrup,worth medicines pvt ltd,80,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,1.000000
33118,COF-RYL Cough Syrup,cipla ltd,82,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,1.000000
36420,Cofryl Syrup,cipla ltd,82,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,1.000000
32725,Cofsils Cough Syrup,cipla health ltd,95,bottle of 100 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,1.000000
199137,Suncaryl Syrup,suncare formulations pvt ltd,110,bottle of 150 ml syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,1.000000


### Embeddings

In [5]:
from sentence_transformers import SentenceTransformer

c:\ML Projects\AIPharma\.venv\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [13]:
dt.head(5)

,id,name,price,Is_discontinued,manufacturer_name,type,pack_size_label,medicine_desc,side_effects,drug_interactions,composition,cleaned_name,document,similarity
0,1,Augmentin 625 Duo Tablet,223,False,glaxo smithkline pharmaceuticals ltd,allopathy,strip of 10 tablets,Augmentin 625 Duo Tablet is a penicillin-type ...,"Vomiting,Nausea,Diarrhea","{""drug"": [], ""brand"": [], ""effect"": []}",amoxycillin (500mg) clavulanic acid (125mg),augmentin 625 duo tablet,"\n Medicine : augmentin 625 duo tablet,\n ...",0.000000
1,2,Azithral 500 Tablet,132,False,alembic pharmaceuticals ltd,allopathy,strip of 5 tablets,Azithral 500 Tablet is an antibiotic used to t...,"Vomiting,Nausea,Abdominal pain,Diarrhea","{""drug"": [""Astemizole"", ""Ebastine"", ""Mizolasti...",azithromycin (500mg),azithral 500 tablet,"\n Medicine : azithral 500 tablet,\n Com...",0.000000
2,3,Ascoril LS Syrup,118,False,glenmark pharmaceuticals ltd,allopathy,bottle of 100 ml syrup,Ascoril LS Syrup is a combination medicine use...,"Nausea,Vomiting,Diarrhea,Upset stomach,Stomach...","{""drug"": [], ""brand"": [], ""effect"": []}",ambroxol (30mg/5ml) levosalbutamol (1mg/5ml),ascoril ls syrup,"\n Medicine : ascoril ls syrup,\n Compos...",0.229844
3,4,Allegra 120mg Tablet,218,False,sanofi india ltd,allopathy,strip of 10 tablets,Allegra 120mg Tablet is an anti-allergy medici...,"Headache,Drowsiness,Dizziness,Nausea","{""drug"": [""Aprepitant"", ""Fosaprepitant"", ""Maga...",fexofenadine (120mg),allegra 120mg tablet,"\n Medicine : allegra 120mg tablet,\n Co...",0.000000
4,5,Avil 25 Tablet,10,False,sanofi india ltd,allopathy,strip of 15 tablets,Avil 25 Tablet is an antiallergic medication u...,Sleepiness,"{""drug"": [""Zafirlukast"", ""Alprazolam"", ""Betahi...",pheniramine (25mg),avil 25 tablet,"\n Medicine : avil 25 tablet,\n Composit...",0.000000


In [6]:
model = SentenceTransformer("all-MiniLM-L6-v2")

Loading weights: 100%|██████████| 103/103 [00:00<00:00, 1080.40it/s, Materializing param=pooler.dense.weight]                             
BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


In [16]:
embeddings = model.encode(
    dt['document'].tolist(),
    show_progress_bar=True,
    batch_size=64
)

Batches: 100%|██████████| 3807/3807 [25:16<00:00,  2.51it/s]


In [ ]:
import numpy as np
import faiss

new_embeddings = np.array(embeddings).astype("float32")
faiss.normalize_L2(new_embeddings)  # Normalize the vectors

#### Create Flat Index (Inner Product)

In [30]:
dimension = new_embeddings.shape[1]

index = faiss.IndexFlatIP(dimension)
index.add(new_embeddings)

#### Save Index 

In [ ]:
faiss.write_index(index, "../data/medicine_cosine_flat.index") 
dt.to_pickle("../data/medicine_metadata.pkl")

#### To load it later run this
>index = faiss.read_index("../Data/medicine_cosine_flat.index")

>df = pd.read_pickle("../Data/medicine_metadata.pkl")

In [35]:
def semantic_search(query, top_k=5):

    query_embedding = model.encode([query]).astype("float32")

    # normalize query
    faiss.normalize_L2(query_embedding)

    scores, indices = index.search(query_embedding, top_k)

    results = dt.iloc[indices[0]].copy()
    results["similarity"] = scores[0]

    return results[
        ["name", "composition", "manufacturer_name", "price", "similarity"]
    ]

In [38]:
semantic_search("benadryl syrup")

,name,composition,manufacturer_name,price,similarity
24840,Benadryl Syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,johnson & johnson ltd,129,0.748119
25019,Benadryl DR Kids Syrup,dextromethorphan hydrobromide (7.5mg/5ml),johnson & johnson ltd,93,0.685259
24855,Benadryl DR Syrup,dextromethorphan hydrobromide (15mg/5ml),johnson & johnson ltd,123,0.662237
103810,Intadryl Syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,intas pharmaceuticals ltd,38,0.607399
16215,Abidryl-A 1.25mg/15mg/50mg/2.5mg Syrup,terbutaline (1.25mg) ambroxol (15mg),abn drugs,49,0.600239


##### Embedding models naturally give too much importance to brand language but in pharma, Composition = Ground Truth

### Core Idea — Dual-Channel Embeddings

---

### Channel 1 — Brand Semantics

Captures how users speak.

Used for:

- brand names

- natural queries

- conversational search

Example : ***augmentin 625 antibiotic tablet by gsk***

---

#### Channel 2 — Composition Semantics (MOST IMPORTANT)

Captures pharmacological equivalence.

Used for:

- substitute detection

- ingredient similarity

- generic matching

Example: ***active ingredients amoxicillin clavulanate potassium antibiotic combination***

---

In [42]:
def build_brand_doc(row):
    return f"""
    Medicine: {row['cleaned_name']},
    Manufacturer: {row['manufacturer_name']},
    Type: {row['cleaned_name']},
    """

def build_composition_doc(row):
    return f"""
    Active ingredients: {row['composition']}
    Pharmaceutical drug composition.
    """

In [ ]:
dt["brand_text"] = dt.apply(build_brand_doc, axis=1)
dt["composition_text"] = dt.apply(build_composition_doc, axis=1)


NameError: name 'dt' is not defined

In [44]:
from sentence_transformers import SentenceTransformer

model = SentenceTransformer("all-MiniLM-L6-v2")

brand_embeddings = model.encode(
    dt["brand_text"].tolist(),
    batch_size=64,
    show_progress_bar=True
)

composition_embeddings = model.encode(
    dt["composition_text"].tolist(),
    batch_size=64,
    show_progress_bar=True
)

Loading weights: 100%|██████████| 103/103 [00:00<00:00, 743.56it/s, Materializing param=pooler.dense.weight]                             
BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
Batches: 100%|██████████| 3807/3807 [09:57<00:00,  6.37it/s]


In [ ]:
import faiss
import numpy as np

dt = pd.read_pickle("../Data/medicine_metadata.pkl")

brand_embeddings = np.array(brand_embeddings).astype("float32")
composition_embeddings = np.array(composition_embeddings).astype("float32")

faiss.normalize_L2(brand_embeddings)
faiss.normalize_L2(composition_embeddings)

TypeError: float() argument must be a string or a real number, not 'IndexFlatIP'

In [49]:
dimension = brand_embeddings.shape[1]

brand_index = faiss.IndexFlatIP(dimension)
brand_index.add(brand_embeddings)

composition_index = faiss.IndexFlatIP(dimension)
composition_index.add(composition_embeddings)

In [50]:
import os

# Define the data directory
DATA_DIR = "../Data"

# 1. Save the Indices
faiss.write_index(brand_index, os.path.join(DATA_DIR, "brand_search.index"))
faiss.write_index(composition_index, os.path.join(DATA_DIR, "composition_search.index"))

# 2. Save the single Metadata file
dt.to_pickle(os.path.join(DATA_DIR, "medicine_metadata_dual.pkl"))

In [12]:
import faiss

brand_index = faiss.read_index("../Data/brand_search.index")
composition_index = faiss.read_index("../Data/composition_search.index")

In [20]:
def find_anchor(query):
    q = model.encode([query]).astype("float32")
    faiss.normalize_L2(q)

    scores, indices = brand_index.search(q, 1)
    return dt.iloc[indices[0][0]]

In [ ]:
def hybrid_search(query, top_k=10, alpha=0.4):
    """
    alpha controls importance:
    alpha small → composition dominates
    alpha large → brand dominates
    """

    query_embedding = model.encode([query]).astype("float32")
    faiss.normalize_L2(query_embedding)

    brand_scores, brand_ids = brand_index.search(query_embedding, top_k)
    comp_scores, comp_ids = composition_index.search(query_embedding, top_k)

    scores = {}

    # brand contribution
    for score, idx in zip(brand_scores[0], brand_ids[0]):
        scores[idx] = scores.get(idx, 0) + alpha * score

    # composition contribution
    for score, idx in zip(comp_scores[0], comp_ids[0]):
        scores[idx] = scores.get(idx, 0) + (1 - alpha) * score

    ranked = sorted(scores.items(), key=lambda x: x[1], reverse=True)

    final_indices = [idx for idx, _ in ranked[:top_k]]

    return dt.iloc[final_indices][
        ["name", "composition", "manufacturer_name", "price"]
    ]

In [ ]:
def find_substitute(query, top_k=10):
    """
    Find substitutes based primarily on composition similarity.

    alpha:
        small  → composition dominates
        large  → brand similarity dominates (optional future use)
    """


    anchor = find_anchor(query)

    if anchor is None:
        return None

    anchor_idx = anchor.name  
    anchor_composition = anchor["composition"]

    comp_text = f"active ingredients {anchor_composition} pharmaceutical drug composition"

    comp_emb = model.encode(
        [comp_text],
        normalize_embeddings=True
    ).astype("float32")

    comp_scores, comp_ids = composition_index.search(comp_emb, top_k + 1)

    scores = {}

    for score, idx in zip(comp_scores[0], comp_ids[0]):
        if idx == anchor_idx:
            continue  # remove same medicine

        scores[idx] = float(score)

    ranked = sorted(scores.items(), key=lambda x: x[1], reverse=True)

    final_indices = [idx for idx, _ in ranked[:top_k]]

    return dt.iloc[final_indices][
        ["name", "composition", "manufacturer_name", "price"]
    ]

In [31]:
find_substitute("benadryl", alpha=1)

,name,composition,manufacturer_name,price
207575,Tussberry-N Syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,acinta pharmaceuticals pvt ltd,51
199137,Suncaryl Syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,suncare formulations pvt ltd,110
197356,Super Tus Syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,worth medicines pvt ltd,80
193007,Suntuss CF Syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,suncare formulations pvt ltd,70
156341,Odaril Syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,moraceae pharmaceuticals pvt ltd,43
145633,New Brethese Syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,galpha laboratories ltd,68
103810,Intadryl Syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,intas pharmaceuticals ltd,38
36420,Cofryl Syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,cipla ltd,82
33118,COF-RYL Cough Syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,cipla ltd,82
32725,Cofsils Cough Syrup,diphenhydramine (14.08mg/5ml) ammonium chlori...,cipla health ltd,95


In [33]:
hybrid_search("composition having diphenhydramine", top_k=10)

,name,composition,manufacturer_name,price
27364,Bedryl 75mg Soft Gelatin Capsule,diphenhydramine (75mg),sun life sciences pvt ltd,50
237111,Zendryl Syrup,diphenhydramine (na),lupin ltd,13
197143,StayHappi Diphenhydramine+Ammonium Chloride+So...,diphenhydramine (14.08mg) ammonium chloride (...,sarvagunaushdhi pvt ltd,36
73272,Exylin 250 DT Tablet,diphenhydramine (250mg),win-medicare pvt ltd,43
59526,Dipren 0.025% Lotion,beclometasone (0.025% w/v),dan laboratories,60
65137,Diphen EX Syrup,ammonium chloride (138mg/5ml) sodium citrate ...,p & b pharmaceuticals ltd,106
60517,Dipin AT 5mg/50mg Tablet,amlodipine (5mg) atenolol (50mg),parangat organics,29
56289,Dipgenta Plus Cream,clobetasol (0.05% w/w) neomycin (0.5% w/w),fulford india ltd,105
58687,Dipnate Plus Lotion,betamethasone (0.05% w/w) zinc sulfate (0.5% ...,karlin pharmaceuticals & exports pvt ltd,75
65856,Dipnate Cream,betamethasone (0.05% w/w),karlin pharmaceuticals & exports pvt ltd,32


Adding Intent Routing

In [34]:
INTENTS = [
    "substitute",
    "price_search",
    "comparison",
    "general_search"
]

intent_examples = {
    "substitute": [
        "similar medicine",
        "alternative drug",
        "replacement for medicine",
        "same composition"
    ],
    "price_search": [
        "cheap medicine",
        "low cost drug",
        "affordable option",
        "under budget"
    ],
    "comparison": [
        "compare medicines",
        "difference between drugs",
        "which is better"
    ],
    "general_search": [
        "medicine information",
        "drug details",
        "what is this medicine"
    ]
}

In [35]:
intent_embeddings = {}

for intent, examples in intent_examples.items():
    emb = model.encode(examples)
    intent_embeddings[intent] = emb.mean(axis=0)

In [36]:
def detect_intent(query):

    q = model.encode([query]).astype("float32")
    faiss.normalize_L2(q)

    best_intent = None
    best_score = -1

    for intent, emb in intent_embeddings.items():
        emb = emb.reshape(1, -1).astype("float32")
        faiss.normalize_L2(emb)

        score = (q @ emb.T)[0][0]

        if score > best_score:
            best_score = score
            best_intent = intent

    return best_intent

In [ ]:
def handle_query(query):

    intent = detect_intent(query)

    if intent == "substitute":
        return find_substitute(query)

    return hybrid_search(query)